### ISSUE TITLE
improper_ctypes incorrectly warns on projections featuring opaque types

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#73251]
cc @davidtwco 

```rust
#![feature(type_alias_impl_trait)]

pub trait Foo {
    type Assoc;
}

impl Foo for () {
    type Assoc = u32;
}

type Bar = impl Foo;

fn assign() -> Bar {}

extern "C" {
    pub fn lint_me() -> <Bar as Foo>::Assoc;
}
```
Gives the following incorrect warning:
```
warning: `extern` block uses type `Bar`, which is not FFI-safe
  --> src/lib.rs:16:25
   |
16 |     pub fn lint_me() -> <Bar as Foo>::Assoc;
   |                         ^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: `#[warn(improper_ctypes)]` on by default
   = note: opaque types have no C equivalent

warning: 1 warning emitted
```

**Comment by davidtwco** [comment]
I'm not sure if this is incorrect - I think this is what I'd expect. In the declaration of `lint_me`, we don't know what concrete type `Bar` is - if it were `()`, then we could normalize to `u32` - but we don't know? Maybe I'm missing something.

**Comment by lcnr** [comment]
Hmm, that might be true. I definitely think that the following should not lint here (with `Assoc` bound to u32)
```rust
#![feature(type_alias_impl_trait)]

pub trait Foo {
    type Assoc;
}

impl Foo for () {
    type Assoc = u32;
}

type Bar = impl Foo<Assoc = u32>;

fn assign() -> Bar {}

extern "C" {
    pub fn lint_me() -> <Bar as Foo>::Assoc;
}

```

**Comment by davidtwco** [comment]
I agree that seems like it should be fine. So, this happens because of this call:

https://github.com/rust-lang/rust/blob/a37c32e2d5fb186627ffe99a391c7fd6fd159334/src/librustc_lint/types.rs#L962-L967

It makes sense that we want to lint on opaque types on extern declarations, it's unclear whether we would want to allow projections which contain opaque types - this would solve the case you've shown here, but could be used to "sneak" opaque types by the lint (in [this playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4ca4b980a93f989bb8bdc4c4b2a9e545), the projection normalizes to an opaque type; it's hard to detect because the normalization goes further and resolves it to `u32` so we can't tell that it was an opaque type) - I have a build with this locally.

Any thoughts on whether we should change the behavior here?

cc @hanna-kruppe

**Comment by lcnr** [comment]
Hm, I think it might be worth it to normalize using `Reveal::UserFacing` here, which should not normalize the first case and your example while fixing the second one if I am not mistaken.

**Comment by hanna-kruppe** [comment]
Long ago I filed https://github.com/rust-lang/rust/issues/60855 where we also concluded Reveal::UserFacing is the right behavior for normalization in this lint. The current behavior comes from #64359 cc @varkor

**Comment by davidtwco** [comment]
I'll put up a PR with my experiments from last night later today in that case.
